### PR TITLE
[WIP; DONTMERGE] Implement #1436

### PR DESCRIPTION
**Pull request by gfgtdf**
Implements #1436

Not complete yet, do not merge it. 

This adds support to use lua and formula in abilitites attributes. The original suggestion in #1436 was to allow variable subsitution but i decided against that since variable substitution in objects (in which abiltites often appear) is a rather messy thing and in most cases one would use formula anyways.

### CODE REVIEW

File: source_lists/wesnoth [hunk 1]
```diff
@@ -370,7 +371,6 @@ team.cpp
 teambuilder.cpp
 terrain/filter.cpp
 tod_manager.cpp
-units/abilities.cpp
```
Comment (CelticMinstrel) [inline source_lists/wesnoth]: Uh, what's with this? Probably revert it?

File: src/map_command_handler.hpp [hunk 2]
```diff
@@ -360,17 +360,13 @@ class map_command_handler
...
+
 	virtual void register_alias(const std::string& to_cmd,
 		const std::string& cmd)
 	{
-		// disable the assert to allow alias to "command + args"
-		// the fonction assert_existence seems unused now
-		//assert_existence(to_cmd);
 		command_alias_map_[cmd] = to_cmd;
 	}
+
```
Comment (CelticMinstrel) [inline src/map_command_handler.hpp]: Does this even have anything to do with the PR?

File: src/menu_events.cpp [hunk 3]
```diff
@@ -1041,11 +1041,6 @@ class console_handler : public map_command_handler<console_handler>, private cha
 		chmap::register_command(cmd, h, help, usage, flags + "N"); // add chat commands as network_only
 	}
 
-	virtual void assert_existence(const std::string& cmd)
-	{
-		chmap::assert_existence(cmd);
-	}
-
```
Comment (CelticMinstrel) [inline src/menu_events.cpp]: Similarly, what does this have to do with the PR?

File: src/units/abilities.cpp [hunk 4]
```diff
@@ -402,39 +408,68 @@ bool unit::has_ability_type(const std::string& ability) const
...
-			if (value < 0) value = -value;
-			if (only_cumulative && value >= abs_max) {
-				abs_max = value;
-				best_loc = p.second;
-			}
-		} else if (only_cumulative || value > flat) {
-			only_cumulative = false;
-			flat = value;
-			best_loc = p.second;
+		utils::string_view formula_pefix("formula:");
```
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Typo - should be `formula_prefix`

File: src/units/abilities.cpp [hunk 5]
```diff
@@ -402,39 +408,68 @@ bool unit::has_ability_type(const std::string& ability) const
...
+				if(u_itor != resources::units->end()) {
+					callable.add("sender", wfl::variant(std::make_shared<wfl::unit_callable>(*u_itor)));
+				}
+				u_itor = resources::units->find(receiver_loc);
+				if(u_itor != resources::units->end()) {
+					callable.add("receiver", wfl::variant(std::make_shared<wfl::unit_callable>(*u_itor)));
+				}
+				return formula_handler(wfl::formula(s), callable);
+			} catch(wfl::formula_error& e) {
+				lg::wml_error() << "Formula error in unit filter: " << e.type << " at " << e.filename << ':' << e.line << ")\n";
```
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: This isn't a unit filter.

File: src/units/abilities.cpp [hunk 6]
```diff
@@ -448,15 +483,18 @@ std::pair<int,map_location> unit_ability_list::lowest(const std::string& key, in
...
 		if ((*p.first)["cumulative"].to_bool()) {
 			stack += value;
 			if (value < 0) value = -value;
-			if (only_cumulative && value <= abs_max) {
+			if (only_cumulative && !comp(value, abs_max)) {
 				abs_max = value;
 				best_loc = p.second;
 			}
-		} else if (only_cumulative || value < flat) {
+		} else if (only_cumulative || comp(flat, value)) {
```
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Suggestion: Consider `comp` to be a less-than operator. In other words, swap the parameters in both calls (and adjust the areas where the comparators are created as well).

This is more consistent with standard library behaviour.

**Comment by CelticMinstrel** [comment]
I don't like the variable names `sender` and `receiver`. What does this even mean?

File: src/units/abilities.cpp (hunk 6 above)
Comment (gfgtdf) [inline src/units/abilities.cpp]: comp is a less-than operator

File: source_lists/wesnoth (hunk 1 above)
Comment (gfgtdf) [inline source_lists/wesnoth]: as said the commit message. this is to make travis fail faster if a do something wrong during the pr (i don't build it locally)

**Comment by gfgtdf** [comment]
the unit has the abbility and the other is effected by it. (think of [leadership], or [damage] effect_opponent=true, etc.) i don'T like them muhc eigher but i couldn't think of better names.

File: src/map_command_handler.hpp (hunk 2 above)
Comment (gfgtdf) [inline src/map_command_handler.hpp]: this commit just happened to be on the branch when i started working on this.

File: source_lists/wesnoth (hunk 1 above)
Comment (CelticMinstrel) [inline source_lists/wesnoth]: Ahh, got it.

File: src/map_command_handler.hpp (hunk 2 above)
Comment (CelticMinstrel) [inline src/map_command_handler.hpp]: I wouldn't squash this in with the rest, at least,

**Comment by CelticMinstrel** [comment]
Is there anything wrong with `unit` and `other_unit`? That would match the situation in standard unit filters. (Standard unit filters also make `unit` the default fallback, so that eg `x` is equivalent to `unit.x`.)

File: src/units/abilities.cpp [hunk 7]
```diff
@@ -402,39 +408,68 @@ bool unit::has_ability_type(const std::string& ability) const
...
-			if (value < 0) value = -value;
-			if (only_cumulative && value >= abs_max) {
-				abs_max = value;
-				best_loc = p.second;
-			}
-		} else if (only_cumulative || value > flat) {
-			only_cumulative = false;
-			flat = value;
-			best_loc = p.second;
+		utils::string_view formula_prefix("formula:");
```
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Okay, so I don't like this way of detecting a formula. Why don't you use the existing `typed_formula` class for values that could optionally be a formula?

File: src/units/abilities.cpp (hunk 6 above)
Comment (gfgtdf) [inline src/units/abilities.cpp]: this diffs shows the difference from the old 'lowest' function to the new 'highest_impl' function. 'highest_impl' is designed to works as 'highest' if the usual default argument  (std::less is given)

**Comment by gfgtdf** [comment]
Unlike for attacks or filters, i think it's too unclear which one is 'primary' unit in this context.

File: src/units/abilities.cpp (hunk 7 above)
Comment (gfgtdf) [inline src/units/abilities.cpp]: the typed_formula class, at lest in its current form involves some overhead: In the most common case the attribute values (integers) is converted to a string and then converted back for no reason. Also It wouldn't allow me to add lua support later.

File: src/units/abilities.cpp (hunk 6 above)
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Ah, alright then. I think I'd recommend changing the name though, since it's no longer a "highest" or "lowest". Maybe something like "get_extremum"? Since "extremum" is the generalization of "minimum" and "maximum".

File: src/units/abilities.cpp (hunk 7 above)
Comment (CelticMinstrel) [inline src/units/abilities.cpp]: I don't think the first complaint really matters - if you really think it's a problem, it can be fixed in  `typed_formula`, anyway.

As for the second complaint... no other place provides WFL and Lua support with the same syntax, so I don't see any particular reason why this one should be any different. Why not use a `lua_function` key for Lua support? Or even a `[lua]` tag?

**Comment by CelticMinstrel** [comment]
I think it's pretty clear that the unit that owns the ability should be the primary unit... but if you're not convinced, maybe something like `owning_unit` and `other_unit`?

File: src/units/abilities.cpp (hunk 7 above)
Comment (gfgtdf) [inline src/units/abilities.cpp]: > I don't think the first complaint really matters

The types_formula objects in gui2 are afaik only constructed once whiel the dialog is created so it's probably not that impotant there, i don't know thought how often this code here is called, so i'd rather try to keep is fast, and wit the current implmentation it's just as fast as it was before.

> Why not use a `lua_function` key for Lua support? Or even a `[lua]` tag?

abilities support many attributes, ofc i could add add_lua_function, sub_lua_function, min_value_lua_function etc attributs but i liek this one better.

Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Uh, you really don't need formula/Lua support for add or sub. They're only needed because we _don't_ have formula/Lua support. If you were using formulas or Lua, you'd just use value instead. For example, add=5 becomes value="(base_value + 5)".

Furthermore, you don't necessarily need a `lua_function` for every attribute, because Lua functions can return multiple values. You could have one `lua_function` that returns _all_ the required attributes.

Please don't use bad syntax just because of concerns of speed. Even if you don't want to use `typed_formula`, I'd much prefer if you'd at least use the same method of recognizing that it's a formula - that is, starting with an open parenthesis makes it a formula.

Comment (gfgtdf) [inline src/units/abilities.cpp]: > Uh, you really don't need formula/Lua support for add or sub. They're only needed because we don't have formula/Lua support. If you were using formulas or Lua, you'd just use value instead. For example, add=5 becomes value="(base_value + 5)".


That's was actuall my first plan, but it seems like add= sub= stc. are handled in a speicifc way that makes it indepnednt on the order in which the abilities appear, so value="(base_value + 5)". will behave diffrent than add=5 if multiple weapon specialy are invloved. (to abilitites with `add=5` will sum up to add=10 while for two abilitites with both `value="(base_value + 5)"` the game will take the higher oen which is still 5)


> Furthermore, you don't necessarily need a lua_function for every attribute, because Lua functions can return multiple values

This probably requires me to write this code in a case by case (where case is a aspeific ability) basis.

> Please don't use bad syntax just because of concerns of speed

speedwise it wouldn't make a difference between a `formula:` and a `(` prefix. I just

Comment (CelticMinstrel) [inline src/units/abilities.cpp]: I think the way to make that still work that way would be for `base_value` to be the result of applying the previous ability... not entirely sure if that can work the same though.

> speedwise it wouldn't make a difference between a formula: and a ( prefix. I just

Forgot to finish your sentence?

Comment (CelticMinstrel) [inline src/units/abilities.cpp]: Using `(` as the prefix also means you don't need a call to `substr` two lines down, by the way.

**Comment by CelticMinstrel** [comment]
I'm considering merging this in with PR 1806, and possibly adding a few tweaks (mainly related to my comments earlier). Any objections?

**Comment by CelticMinstrel** [comment]
Feel free to also cherry-pick 9bae92a, since I didn't merge that in.
